discovery-provider/src/tasks/karma.py

```diff
@@ -0,0 +1,64 @@
+import logging
+import time
+from src.models import Track
+from src.tasks.celery_app import celery
+from src.queries.get_trending_tracks import make_trending_cache_key, generate_unpopulated_trending
+from src.utils.redis_cache import pickle_and_set
+
+logger = logging.getLogger(__name__)
+time_ranges = ["week", "month", "year"]
+
+def get_genres(session):
+    """Returns all genres"""
+    genres = (
+        session.query(
+            Track.genre
+        ).distinct(
+            Track.genre
+        )).all()
+    genres = filter(lambda x: x[0] is not None and x[0] != "", genres)
+    return list(map(lambda x: x[0], genres))
+
+
+def update_karma(self, db, redis):
+    logger.info('Running update karma')
+    update_start = time.time()
+    with db.scoped_session() as session:
+        genres = get_genres(session)
+
+        # Make sure to cache empty genre
+        genres.append(None)
+
+        for genre in genres:
+            for time_range in time_ranges:
+                cache_start_time = time.time()
+                res = generate_unpopulated_trending(session, genre, time_range)
+                key = make_trending_cache_key(time_range, genre)
+                pickle_and_set(redis, key, res)
+                cache_end_time = time.time()
+                total_time = cache_end_time - cache_start_time
+                logger.info(f"Cached trending for {genre}-{time_range} in {total_time} seconds")
+    update_end = time.time()
+    update_total = update_end - update_start
+    logger.info(f"Finished update karma in {update_total} seconds")
+
+######## CELERY TASKS ########
+@celery.task(name="update_karma", bind=True)
+def update_karma_task(self):
+    """Caches all trending combination of time-range and genre (including no genre)."""
+    db = update_karma_task.db
+    redis = update_karma_task.redis
+    have_lock = False
+    update_lock = redis.lock("karma_lock", timeout=7200)
```

Comment:
i always get this confused and need to look this up. but does timeout refer to the the TTL of the lock itself? thats what i gather from here https://redis-py.readthedocs.io/en/stable/#redis.Redis.lock

Reply:
this may be a broader config to address, but seeing as how the task runs every hour should be okay?

Reply:
Yeah - this was a copy paste and I don't think it does anything for us, but shouldn't hurt us either. If we have a job running long enough for this to cause an issue we have other problems